lib/supavisor/protocol/client.ex

```diff
@@ -0,0 +1,157 @@
+defmodule Supavisor.Protocol.Client do
+  require Logger
+
+  @pkt_header_size 5
+
+  defmodule(Pkt,
+    do: defstruct([:tag, :len, :payload, :bin])
+  )
+
+  def pkt_header_size, do: @pkt_header_size
+
+  def header(<<char::8, pkt_len::32>>) do
+    {tag(char), pkt_len}
+  end
+
+  @spec decode(binary) :: {:ok, [Pkt.t()], binary} | {:error, any}
+  def decode(data) do
+    decode(data, [])
+  end
+
+  @spec decode(binary, [Pkt.t()]) :: {:ok, [Pkt.t()], binary} | {:error, any}
+  def decode("", acc), do: {:ok, Enum.reverse(acc), ""}
+
+  def decode(data, acc) do
+    case decode_pkt(data) do
+      {:ok, pkt, rest} -> decode(rest, [pkt | acc])
+      {:error, :payload_too_small} -> {:ok, Enum.reverse(acc), data}
+    end
+  end
+
+  @spec decode_pkt(binary) ::
+          {:ok, Pkt.t(), binary}
+          | {:acc, nil, binary}
+          | {:error, :payload_too_small}
+  def decode_pkt(<<_::8, pkt_len::32, payload::binary>>)
+      when byte_size(payload) < pkt_len - 4 do
+    {:error, :payload_too_small}
+  end
+
+  def decode_pkt(<<char::8, pkt_len::32, rest::binary>>) do
+    case tag(char) do
+      nil ->
+        {:error, {:undefined_tag, <<char>>}}
+
+      tag ->
+        payload_len = pkt_len - 4
+        <<bin_payload::binary-size(payload_len), rest2::binary>> = rest
+
+        {:ok,
+         %Pkt{
+           tag: tag,
+           len: pkt_len + 1,
+           payload: decode_payload(tag, bin_payload),
```

Comment:
Right now it seems you only use the payload, perhaps it is worth simplifying or do you plan to use further functionality?

Reply:
I wrote this function in the beginning when I was learning pg protocol. I'll change it to something that returns only the payload 👍